feature/manager/src/test/java/com/kinandcarta/create/proxytoggle/manager/viewmodel/ProxyManagerViewModelTest.kt

```diff
@@ -0,0 +1,385 @@
+        }
+
+        dispatcher.scheduler.advanceUntilIdle()
+        subject.uiState.value.apply {
+            assertThat(proxyEnabled).isEqualTo(false)
+            assertThat(addressState.error).isNotNull()
+            assertThat(addressState.forceFocus).isTrue()
+            assertThat(portState.error).isNull()
+            assertThat(portState.forceFocus).isFalse()
+        }
+    }
+
+    @Test
+    fun `toggleProxy() - GIVEN disabled proxy and invalid port THEN show port error`() {
+        // GIVEN
+        every { mockProxyValidator.isValidIP(PROXY_ADDRESS) } returns true
+        every { mockProxyValidator.isValidPort(PROXY_PORT) } returns false
+        givenTextFieldValues(PROXY_ADDRESS, PROXY_PORT)
+
+        // WHEN
+        subject.toggleProxy()
+
+        // THEN
+        verify {
+            mockThemeSwitcher.isNightMode()
+            mockProxyValidator.isValidIP(PROXY_ADDRESS)
+            mockProxyValidator.isValidPort(PROXY_PORT)
+            mockDeviceSettingsManager wasNot Called
+        }
+
+        dispatcher.scheduler.advanceUntilIdle()
+        subject.uiState.value.apply {
+            assertThat(proxyEnabled).isEqualTo(false)
+            assertThat(addressState.error).isNull()
+            assertThat(addressState.forceFocus).isFalse()
+            assertThat(portState.error).isNotNull()
+            assertThat(portState.forceFocus).isTrue()
+        }
+    }
+
+    @Test
+    fun `onAddressChanged() - WHEN address changes THEN update uiState`() {
+        // GIVEN
+        givenTextFieldValues("", PROXY_PORT)
+
+        // WHEN
+        subject.onAddressChanged(PROXY_ADDRESS)
+
+        // THEN
+        verify { mockThemeSwitcher.isNightMode() }
+        assertThat(subject.uiState.value.addressState.text).isEqualTo(PROXY_ADDRESS)
+    }
+
+    @Test
+    fun `onPortChanged() - WHEN port changes THEN update uiState`() {
+        // GIVEN
+        givenTextFieldValues(PROXY_ADDRESS, "")
+
+        // WHEN
+        subject.onPortChanged(PROXY_PORT)
+
+        // THEN
+        verify { mockThemeSwitcher.isNightMode() }
+        assertThat(subject.uiState.value.portState.text).isEqualTo(PROXY_PORT)
+    }
+
+    @Test
+    fun `onAddressChanged() - simple filter, only digits and dots`() {
+        // WHEN
+        val userTyped = "125j0hrvhiz89@$)(!V."
+        subject.onAddressChanged(userTyped)
+
+        // THEN
+        verify { mockThemeSwitcher.isNightMode() }
+        assertThat(subject.uiState.value.addressState.text).isEqualTo(
+            userTyped.filter { it.isDigit() || it == '.' }
+        )
+    }
+
+    @Test
+    fun `onPortChanged() - simple filter, only up to 5 digits`() {
+        // WHEN
+        val userTyped = "125j0hrvhiz89@$)(!V."
+        subject.onPortChanged(userTyped)
+
+        // THEN
+        verify { mockThemeSwitcher.isNightMode() }
+        assertThat(subject.uiState.value.portState.text).isEqualTo(
+            userTyped.filter(Char::isDigit).take(5)
+        )
```

Comment:
```suggestion
        assertThat(subject.uiState.value.portState.text).isEqualTo(12508)
```
Asserting here should be what you expect the exact value.

Reply:
Agreed.